crates/jstzd/src/task/task.rs

```diff
@@ -0,0 +1,253 @@
+use tokio::sync::{mpsc, OnceCell};
+
+#[async_trait]
+pub trait Task {
+    /// A set of tasks which are all executed on the same 'execution backend'
+    type TaskSet;
+
+    /// Performs the unit of work for the task
+    async fn run_task(
+        self,
+        signals: mpsc::Receiver<Signal>,
+        task_group: &Self::TaskSet,
+    ) -> anyhow::Result<()>;
+
+    /// Spawns the task
+    async fn spawn_task(self) -> JoinHandle;
+}
+
+pub struct JoinHandle {
+    task: OnceCell<tokioJoinHandle<()>>,
+    allowed_signal: SignalKind,
+    dropped: bool,
+}
+
+// Need this because SignalKind does not have default
+impl Default for JoinHandle {
+    fn default() -> Self {
+        JoinHandle {
+            task: OnceCell::new(),
+            allowed_signal: SignalKind::terminate(),
+            dropped: false,
+        }
+    }
+}
+
+impl JoinHandle {
+    /// Creates a new join handle from a task.
+    ///
+    /// # Examples
+    ///
+    /// ```rust
+    /// use jstzd::task::task::JoinHandle;
+    /// use tokio::signal::unix::SignalKind;
+    ///
+    /// # tokio_test::block_on(async {
+    /// let handle = JoinHandle::new(
+    ///     tokio::spawn(async move {
+    ///         tokio::time::sleep(std::time::Duration::from_secs(10)).await;
+    ///     }),
+    ///     SignalKind::terminate(),
+    /// );
+    /// # });
+    /// ```
+    pub fn new(task: tokioJoinHandle<()>, signal: SignalKind) -> Self {
+        JoinHandle {
+            task: OnceCell::from(task),
+            allowed_signal: signal,
+            ..Default::default()
+        }
+    }
+
+    /// This function will block until the task has exited, returning a success or an error
+    async fn wait(&mut self) -> anyhow::Result<()> {
+        return match self.task.take() {
+            Some(handle) => {
+                let result = handle.await;
+                if result.is_err() {
+                    let err = result.unwrap_err();
+                    if err.is_cancelled() {
+                        return anyhow::Ok(());
+                    }
+                    return Err(err.into());
+                }
+                anyhow::Ok(())
+            }
+            None => anyhow::Ok(()),
+        };
+    }
+
+    /// This function will signal the running task with the specified signal
+    /// This is a non-blocking operation
+    fn signal(&self, signal: SignalKind) -> anyhow::Result<()> {
+        if signal != self.allowed_signal && signal != SignalKind::terminate() {
+            return Err(anyhow::anyhow!(
+                "unacceptable signal '{:?}'",
+                signal.as_raw_value()
+            ));
+        }
+        if let Some(handle) = self.task.get() {
+            handle.abort();
+        }
+        return anyhow::Ok(());
+    }
+
+    async fn shutdown(&mut self) {
+        let _ = self.signal(SignalKind::terminate());
+        let _ = self.wait().await;
+    }
+}
+
+impl Future for JoinHandle {
+    type Output = ();
```

Comment:
Shouldn't we return an `anyhow::Result<()>`? Since `.await` is essentially `.wait`?

Reply:
Yes. This is now updated.